network/local.go

```diff
@@ -0,0 +1,459 @@
+
+	// outgoing knows how to store packet into the incoming's queue
+	ccc.queues[outEndpoint] = outgoing.connQueue
+	// incoming knows how to store packet into the outgoing's queue
+	ccc.queues[incEndpoint] = incoming.connQueue
+
+	go fn(incoming)
+	return outgoing, nil
+}
+
+// Send will get the connection denoted by this endpoint and will call queueMsg
+// with the packet as argument on it. It returns ErrClosed if it does not find
+// the connection.
+func (ccc *LocalContext) Send(e endpoint, nm Packet) error {
+	ccc.Lock()
+	defer ccc.Unlock()
+	q, ok := ccc.queues[e]
+	if !ok {
+		return ErrClosed
+	}
+
+	q.Push(nm)
+	return nil
+}
+
+// Close will get the connection denoted by this endpoint and will Close it if
+// present.
+func (ccc *LocalContext) Close(conn *LocalConn) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	// delete this conn
+	delete(ccc.queues, conn.local)
+	// and delete the remote one + close it
+	remote, ok := ccc.queues[conn.remote]
+	if !ok {
+		return
+	}
+	delete(ccc.queues, conn.remote)
+	remote.Close()
+}
+
+// Len returns how many local connections is there
+func (ccc *LocalContext) Len() int {
+	ccc.Lock()
+	defer ccc.Unlock()
+	return len(ccc.queues)
+}
+
+// LocalConn is a connection that send and receive messages through channels
+type LocalConn struct {
+	local  endpoint
+	remote endpoint
+
+	// connQueue is the part that is accesible from the LocalContext (i.e. is
+	// shared). Reason why we can't directly share LocalConn is because go test
+	// -race detects as data race (while it's *protected*)
+	*connQueue
+
+	ctx *LocalContext
+}
+
+// newLocalConn simply init the fields of a LocalConn but do not try to
+// connect. It should not be used as-is, most user wants to call NewLocalConn.
+func newLocalConn(ctx *LocalContext, local, remote endpoint) *LocalConn {
+	return &LocalConn{
+		remote:    remote,
+		local:     local,
+		connQueue: newConnQueue(),
+		ctx:       ctx,
+	}
+}
+
+// NewLocalConn returns a new channel connection from local to remote
+// Mimics the behavior of NewTCPConn => tries connecting right away.
+// It uses the default local context.
+func NewLocalConn(local, remote Address) (*LocalConn, error) {
+	return NewLocalConnWithContext(defaultLocalContext, local, remote)
+}
+
+// NewLocalConnWithContext is similar to NewLocalConn but takes a specific
+// LocalContext.
+func NewLocalConnWithContext(ctx *LocalContext, local, remote Address) (*LocalConn, error) {
+	return ctx.Connect(local, remote)
+}
+
+// Send implements the Conn interface.
+func (cc LocalConn) Send(ctx context.Context, msg Body) error {
+
+	var body Body
+	var val = reflect.ValueOf(msg)
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+	body = val.Interface()
+
+	var typ = TypeFromData(body)
+	nm := Packet{
+		MsgType: typ,
+		Msg:     body,
+	}
+	return cc.ctx.Send(cc.remote, nm)
+}
+
+// Receive implements the Conn interface.
+func (cc *LocalConn) Receive(ctx context.Context) (Packet, error) {
```

Comment:
IIRC originally we wanted to use the context to pass data/state from/to different layers of the framework. We don't do that at all for the moment. Not sure if we should just remove the context as currently it isn't really used (I guess yes).